Allow from 0 to 5 preheat setups (bugfix)

GMagician opened the pull request:
- this will complete the "more than 2" preheats setup integration
- added possibility to have 0 or 1 preheats setup (freeing up memory)

Fix also #18450

GMagician committed:
Allow from 0 to 5 preheats setup

This will complete "more than 2" preheats integration
Fix also #18450

GMagician committed:
Allow from 0 to 5 preheats setup

This will complete "more than 2" preheats integration
Fix also #18450

GMagician committed:
Fix Japanese typo

GMagician committed:
Fix Japanese typo

thinkyhead committed:
Update language_an.h

thinkyhead committed:
Update language_an.h

thinkyhead committed:
Update language_bg.h

thinkyhead committed:
Update language_bg.h

thinkyhead committed:
Fix spacing with regex

thinkyhead committed:
Fix spacing with regex

thinkyhead committed:
And more strings

thinkyhead committed:
And more strings

thinkyhead commented on Marlin/src/lcd/menu/menu_ubl.cpp:
```diff
@@ -315,37 +333,39 @@ void _lcd_ubl_invalidate() {
+    #else
+      #define BUILD_MESH_GCODE_ITEM(p)  GCODES_ITEM(MSG_UBL_BUILD_MESH_M##p, PSTR(  \
+                                          "G28\n" \
+                                           "M109 S" STRINGIFY(PREHEAT_##p##_TEMP_HOTEND) "\n" \
```
All these preheat items are editable. One of the maintainers of UBL needs to catch up on that.

thinkyhead committed:
Update menu_configuration.cpp

thinkyhead committed:
Update menu_configuration.cpp

thinkyhead committed:
Simplify setting preset temps

Add an I parameter to M104, M106, M109, M140, and M190 so temperatures can be set from presets.

thinkyhead committed:
Simplify setting preset temps

Add an I parameter to M104, M106, M109, M140, and M190 so temperatures can be set from presets.

thinkyhead committed:
Merge branch 'allow-more-than-2-preheat-options' of https://github.com/GMagician/Marlin into pr/18488

thinkyhead committed:
Update comments, preset

thinkyhead committed:
Update comments, preset

thinkyhead committed:
Use indexed presets

thinkyhead committed:
Use indexed presets

thinkyhead committed:
Update language_zh_TW.h

thinkyhead committed:
Update language_zh_TW.h

thinkyhead committed:
Update language_test.h

thinkyhead committed:
Update language_test.h

thinkyhead committed:
Add ultralcd include

thinkyhead committed:
Add ultralcd include

thinkyhead committed:
Add G26 I for material preset

thinkyhead committed:
Add G26 I for material preset

thinkyhead committed:
Fewer boolies

thinkyhead committed:
Fewer boolies

thinkyhead committed:
Fix a typo

thinkyhead committed:
Fix a typo

GMagician committed:
Uniform define argument

GMagician committed:
Uniform define argument

GMagician committed:
enforce base 0 argument

GMagician committed:
enforce base 0 argument

thinkyhead committed:
Update M140_M190.cpp

thinkyhead committed:
Update M140_M190.cpp

thinkyhead commented:
Before this is merged, I suppose it would be good to get rid of the need to define so many preheat strings in the language files, and use substitution instead. The simplest substitution would be to use `S_FMT` which works with the `printf` functions. Alternatively, custom substitution with an extension to the menu items could also work.

The thing is, in future I want users to be able to choose the name of each preset from a larger list of 10 or more materials. So, these fixed language strings will need to be replaced to support that added flexibility anyway.

GMagician commented:
> The simplest substitution would be to use S_FMT which works with the printf functions. Alternatively, custom substitution with an extension to the menu items could also work, via lcd_put_u8str_ind_P.

I'll check how to manage this.

GMagician commented:
> I suppose it would be good to get rid of the need to define so many preheat strings

I checked how menu items work and nothing is ready to accept a dynamic string then I think some sort of new ACTION_ITEM must be created...something ACTION_ITEM_FMT? or completely wrong direction?

thinkyhead commented:
> I'll check how to manage this.

There is currently a set of menu item macros which take an index (e.g., `MENU_ITEM_N_P`). This index value (`MenuItemBase::itemIndex`) is then passed on to `lcd_put_u8str_ind_P` to display a number with an optional "`E`" prefix. The `~` and `*` characters are used for these substitutions. I propose to add another substitution character "`$`" and another optional parameter taking a pointer to a list of PSTR pointers (which itself may be stored in PROGMEM). When `$` is seen, the index value from the menu item can be used as a lookup into the list of PSTR pointers, and the string from the list substituted in place.

So, it means extending `MenuItemBase` with a static pointer to a list of strings (or a genera-use `void` pointer) to be used for this kind of extension. The pointer could be used in some instances for a single string substitution, or to point to some other data that the menu item will use for display.

thinkyhead committed:
String list substitution

thinkyhead committed:
String list substitution

GMagician commented on Marlin/src/inc/Conditionals_post.h:
```diff
@@ -2101,11 +2101,9 @@
   #elif defined(PREHEAT_2_LABEL)
     #define PREHEAT_COUNT 2
-  #else
+  #elif defined(PREHEAT_1_LABEL)
```
This should not be that way...lavel 1 check is already done above

GMagician replied:
ok I saw now...just remove on fisrt if. I'm trying to add same logic to all mesh preheats

GMagician committed:
Remove unused messages

GMagician committed:
Remove unused messages

GMagician commented:
I'm honest, 'String list substitution' commit goes over my Marlin knowledge...

thinkyhead commented:
I left the `MSG_PREHEAT_1` (etc.) language strings in place intentionally, because there is `ExtUI` (and other LCD) code which is far from ready to be adapted to incorporate this string substitution method.

GMagician commented:
> I left the MSG_PREHEAT_1

then I'll revert and keep only configuration_post change

GMagician committed:
Revert "Remove unused messages"

This reverts commit 1936c7f7db0d8e146f2075c62ec5b395ddb1a666.

GMagician committed:
Revert "Remove unused messages"

This reverts commit 1936c7f7db0d8e146f2075c62ec5b395ddb1a666.

GMagician committed:
Remove duplicated if test

GMagician committed:
Remove duplicated if test

GMagician committed:
Update some languages

GMagician committed:
Update some languages

thinkyhead commented:
Here are some of the materials that we are likely to include in the select list for the future when we can choose a material name for each preset…
```cpp
typedef enum MaterialType : uint8_t {
  MATERIAL_PLA,
  MATERIAL_PVA,
  MATERIAL_NYLON,
  MATERIAL_PVC,
  MATERIAL_FLEX,
  MATERIAL_PVC,
  MATERIAL_HTPLA,
  MATERIAL_PETG,
  MATERIAL_ABS,
  MATERIAL_ASA,
  MATERIAL_HIPS,
  MATERIAL_PC,
  MATERIAL_PEEK,
  //MATERIAL_CUSTOM1,
  //MATERIAL_CUSTOM2,
  //MATERIAL_CUSTOM3,
  //MATERIAL_CUSTOM4,
  MATERIAL_COUNT
} material_t;
```

GMagician commented:
I was going to test this but I found that it can't be complied due to filament runout change (and not completed at time I made this).
Checking code I suspect now it works in reverse way. Il post a issue to get confirm

GMagician commented:
I had to rebase it to compile and test. It seems not working it shows a $ instead of PLA/PETG, investigating

GMagician committed:
Fix new messages display

thinkyhead committed:
Simpler array

thinkyhead committed:
Merge commit '301e05b' into pr/18488

thinkyhead committed:
Prevent endless loop, add continue

thinkyhead committed:
Patch and clean up macros

thinkyhead committed:
Move fonts info to lcdprint.h, with dummy

thinkyhead committed:
char char center char

thinkyhead committed:
More macro cleanup

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/18488

thinkyhead commented:
I just did a test with the code in its current state, and it worked well for me. I think this can be merged for further ~~mayhem~~ testing and exercise.

GMagician commented on Marlin/src/lcd/lcdprint.cpp:
```diff
@@ -57,10 +57,8 @@ lcd_uint_t lcd_put_u8str_ind_P(PGM_P const pstr, const int8_t ind, PGM_P const i
-    else if (ch == '$') {
+    else if (ch == '$')
       n -= lcd_put_u8str_max_P(inStr, n * (MENU_FONT_WIDTH)) / (MENU_FONT_WIDTH);
-      continue;
```
That's not good, without continue it will print $ as well

thisiskeithb commented:
@thinkyhead: The only issue I found was with 2004/character LCDs displaying a percent symbol after the right arrow (looks fine on a 12864/full graphic LCD). I dug through the related code and couldn't find an obvious reason for it:

![image](https://user-images.githubusercontent.com/13375512/87203497-93a3a300-c2b7-11ea-9bb2-e54b6c7e0e4b.png)

GMagician commented:
it seems writing 1 char less. In my display '>' is on last column

thisiskeithb commented:
> it seems writing 1 char less. In my display '>' is on last column

Which LCD do you have? It looks fine with a 12864/full graphic controller:
![image](https://user-images.githubusercontent.com/13375512/87206725-4e836f00-c2bf-11ea-9fcd-dca66be700d9.png)

thisiskeithb commented:
It repeats the last character from the menu item above it on the 2004 LCDs. Here's a photo from the Configuration menu:
![image](https://user-images.githubusercontent.com/13375512/87208298-19791b80-c2c3-11ea-9177-325f4320d400.png)

GMagician commented:
> Which LCD do you have? It looks fine with a 12864/full graphic controller:

Yes this is what I have

thisiskeithb commented:
> Yes this is what I have

I dug through the code, but couldn't find an obvious reason for the error on 2004/character LCDs. Have any ideas?

GMagician commented:
Something has been changed on the for cycle, but this shouldn't "repeat" last char of previous line

thinkyhead commented:
> Something has been changed on the for cycle, but this shouldn't "repeat" last char of previous line

It's just not writing over it as it scrolls.
